addons/website_crm_score/models/sales_team.py

```diff
@@ -0,0 +1,256 @@
+                    for lead in leads:
+                        if lead.id not in leads_done:
+                            leads_duplicated = lead.get_duplicated_leads(False)
+                            if len(leads_duplicated) > 1:
+                                self.env["crm.lead"].browse(leads_duplicated).merge_opportunity(False, False)
+                            leads_done += leads_duplicated
+                        self._cr.commit()
+                self._cr.commit()
+
+    @api.model
+    def assign_leads_to_salesmen(self, all_team_users, dry=False):
+        users = []
+        for su in all_team_users:
+            if (su.maximum_user_leads - su.leads_count) <= 0:
+                continue
+            domain = safe_eval(su.team_user_domain or '[]', evaluation_context)
+            domain.append(('user_id', '=', False))
+            domain.append(('assign_date', '=', False))
+            domain.append(('score', '>=', su.team_id.min_for_assign))
+
+            # assignation rythm: 2 days of leads if a lot of leads should be assigned
+            limit = int(math.ceil(su.maximum_user_leads / 15.0))
+
+            if dry:
+                dry_leads = self.env["leads.dry.run"].search([('team_id', '=', su.team_id.id)])
+                domain.append(['id', 'in', [dl.lead_id.id for dl in dry_leads]])
+            else:
+                domain.append(('team_id', '=', su.team_id.id))
+
+            leads = self.env["crm.lead"].search(domain, order='score desc', limit=limit * len(su.team_id.team_user_ids))
+            users.append({
+                "su": su,
+                "nbr": min(su.maximum_user_leads - su.leads_count, limit),
+                "leads": leads
+            })
+
+        assigned = {}
+        while users:
+            i = 0
+
+            # statistically select the user that should receive the next lead
+            idx = randint(0, reduce(lambda nbr, x: nbr + x['nbr'], users, 0) - 1)
+
+            while idx > users[i]['nbr']:
+                idx -= users[i]['nbr']
+                i += 1
+            user = users[i]
+
+            # Get the first unassigned leads available for this user
+            while user['leads'] and user['leads'][0] in assigned:
+                user['leads'] = user['leads'][1:]
+            if not user['leads']:
+                del users[i]
+                continue
+
+            #lead convert for this user
+            lead = user['leads'][0]
+            assigned[lead] = True
+            if dry:
+                values = {'lead_id': lead.id, 'team_id': user['su'].team_id.id, 'user_id': user['su'].user_id.id}
+                self.env['leads.dry.run'].create(values)
+            else:
+                # Assign date will be setted by write function
+                data = {'user_id': user['su'].user_id.id}
+                lead.write(data)
+
+                lead.convert_opportunity(lead.partner_id and lead.partner_id.id or None)
+                self._cr.commit()
```

Comment:
idem previous, convert_opp cannot be rollbacked... email and others stuffs are not reversible